cachelib/navy/common/FdpNvme.cpp

```diff
@@ -0,0 +1,321 @@
+}
+
+void FdpNvme::prepFdpUringCmdSqe(struct io_uring_sqe& sqe,
+                                 void* buf,
+                                 size_t size,
+                                 off_t start,
+                                 uint8_t opcode,
+                                 uint8_t dtype,
+                                 uint16_t dspec) {
+  uint32_t maxTfrSize = nvmeData_.getMaxTfrSize();
+  if ((maxTfrSize != 0) && (size > maxTfrSize)) {
+    throw std::invalid_argument("Exceeds max Transfer size");
+  }
+  // Clear the SQE entry to avoid some arbitrary flags being set.
+  memset(&sqe, 0, sizeof(struct io_uring_sqe));
+
+  sqe.fd = file_.fd();
+  sqe.opcode = IORING_OP_URING_CMD;
+  sqe.cmd_op = NVME_URING_CMD_IO;
+
+  struct nvme_uring_cmd* cmd = (struct nvme_uring_cmd*)&sqe.cmd;
+  if (cmd == nullptr) {
+    throw std::invalid_argument("Uring cmd is NULL!");
+  }
+  memset(cmd, 0, sizeof(struct nvme_uring_cmd));
+  cmd->opcode = opcode;
+
+  // start LBA of the IO = Req_start (offset in partition) + Partition_start
+  uint64_t sLba = (start >> nvmeData_.lbaShift()) + nvmeData_.partStartLba();
+  uint32_t nLb = (size >> nvmeData_.lbaShift()) - 1; // nLb is 0 based
+
+  /* cdw10 and cdw11 represent starting lba */
+  cmd->cdw10 = sLba & 0xffffffff;
+  cmd->cdw11 = sLba >> 32;
+  /* cdw12 represent number of lba's for read/write */
+  cmd->cdw12 = (dtype & 0xFF) << 20 | nLb;
+  cmd->cdw13 = (dspec << 16);
+  cmd->addr = (uint64_t)buf;
+  cmd->data_len = size;
+
+  cmd->nsid = nvmeData_.nsId();
+}
+
+void FdpNvme::prepReadUringCmdSqe(struct io_uring_sqe& sqe,
+                                  void* buf,
+                                  size_t size,
+                                  off_t start) {
+  // Placement Handle is not used for read.
+  prepFdpUringCmdSqe(sqe, buf, size, start, nvme_cmd_read, 0, 0);
+}
+
+void FdpNvme::prepWriteUringCmdSqe(
+    struct io_uring_sqe& sqe, void* buf, size_t size, off_t start, int handle) {
+  static constexpr uint8_t kPlacementMode = 2;
+  uint16_t pid;
+
+  if (handle == -1) {
+    pid = getFdpPID(kDefaultPIDIdx); // Use the default stream
+  } else if (handle >= 0 && handle <= maxPIDIdx_) {
+    pid = getFdpPID(static_cast<uint16_t>(handle));
+  } else {
+    throw std::invalid_argument("Invalid placement identifier");
+  }
+
+  prepFdpUringCmdSqe(sqe, buf, size, start, nvme_cmd_write, kPlacementMode,
+                     pid);
+}
+
+// It returns "nvme0n1" for both "/dev/nvme0n1" and "/dev/nvme0n1p1"
+std::string getNvmeBlockName(const std::string& blockDevice) {
+  size_t bPos = blockDevice.find_last_of("/") + 1;
+  size_t pPos = blockDevice.find('p', bPos);
+
+  XDCHECK(bPos < std::string::npos);
+  return blockDevice.substr(bPos, pPos - bPos);
+}
+
+// It returns nvme0n1p1 for /dev/nvme0n1p1 (partitioned NS),
+// and returns "" for /dev/nvme0n1 (non-partitioned NS)
+std::string getNvmePartName(const std::string& blockDevice) {
+  size_t bPos = blockDevice.find_last_of("/") + 1;
```

Comment:
I think you need to check the result of `find_last_of`. `npos + 1` would be 0, which make below code not work as intended.